feat(sns): Implement `RegisterExtension` proposals

aterga opened the pull request:
< [Previous PR](https://github.com/dfinity/ic/pull/5266) |

ShahriarJavidi commented on rs/sns/treasury_manager/src/lib.rs:
```diff
@@ -0,0 +1,184 @@
+    /// Implements the `withdraw` API query function.
+    fn audit_trail(&self, request: AuditTrailRequest) -> AuditTrail;
+
+    /// Implements the `withdraw` API query function.
```
change the comment.

ShahriarJavidi commented on rs/sns/treasury_manager/src/lib.rs:
```diff
@@ -0,0 +1,184 @@
+        let suffix = match op {
+            TreasuryManagerOperation::Deposit => "deposit",
+            TreasuryManagerOperation::Balances => "balances",
+            TreasuryManagerOperation::IssueReward => "reward",
```
NIT: match the suffix with the operation.

ShahriarJavidi commented on rs/sns/treasury_manager/src/lib.rs:
```diff
@@ -0,0 +1,184 @@
+    /// Total amount that may be consumed, including the fees.
+    pub amount_decimals: Nat,
+
+    /// The owner account is used to return the leftover assets and issue rewards.
```
```suggestion
    /// The owner account is used to return the leftover assets and issued rewards.
```

ShahriarJavidi commented on rs/sns/governance/src/pb/conversions.rs:
```diff
@@ -581,6 +581,126 @@ impl From<pb_api::RegisterDappCanisters> for pb::RegisterDappCanisters {
     }
 }
 
+impl From<pb::precise_value::PreciseValue> for pb_api::PreciseValue {
```
As discussed before, we might need to rename one of the `PreciseValue` variables. Here, we have 3 different ones.

aterga commented on rs/sns/governance/api/src/ic_sns_governance.pb.v1.rs:
```diff
@@ -480,6 +480,36 @@ pub struct RegisterDappCanisters {
+#[derive(
+    candid::CandidType, candid::Deserialize, comparable::Comparable, Clone, Debug, PartialEq,
+)]
+pub struct ExtensionInit {
```
Precise value is an enum, so it's slightly safer to have `opt record` at the top level to support API evolution. 

Whereas, for `opt variant`, Candid makes it hard to evolve the API, as even adding new variant cases is risky for the clients, since old clients cannot decode new (extended) variants and fall back to `null`.

aterga commented:
@ShahriarJavidi, this PR got into a weird state, so I've addressed your comments in https://github.com/dfinity/ic/pull/5672 (which is a cleaned up version of this PR). Let's continue the discussion there.
